### PR TITLE
Fix Issue #3: Add Support for Identifying Identical Table Names Across Different Data Sets

### PR DESCRIPTION
**Pull request by KazuSh1geru**
# What is this PR ?

This PR solve https://github.com/devoteamgcloud/dataform-assertions/issues/3

- https://github.com/devoteamgcloud/dataform-assertions/issues/3

# Overview

- Expanded example.js to accept schema.
- Expanded JavaScript in includes/assertions to accept schema.

# Information

## Context Method: `ref`

`ref` method use individual arguments for the "schema", and "name" values.

>References another action, adding it as a dependency to this action, returning valid SQL to be used in a from expression.
This function can be called with a Resolvable object, for example: ${ref({ name: "name", schema: "schema", database: "database" })}

> This function can also be called using individual arguments for the "database", "schema", and "name" values. When only two values are provided, the default database is used and the values are interpreted as "schema" and "name". When only one value is provided, the default database and schema are used, with the provided value interpreted as `"name"`. ${ref("database", "schema", "name")} ${ref("schema", "name")} ${ref("name")}

[document](https://cloud.google.com/dataform/docs/reference/dataform-core-reference#commoncontext)

### CODE REVIEW

File: definitions/example.js [hunk 1]
```diff
@@ -12,52 +12,69 @@ const commonAssertionsResult = commonAssertions({
     // "disabledInEnvs": ["dv", "qa"]
   },
   rowConditions: {
-    "first_table": {
-      "id_not_null": "id IS NOT NULL",
-      "id_strict_positive": "id > 0"
-    },
-    "second_table": {
-      "id_in_accepted_values": "id IN (1, 2, 3)"
+    // Format: "schema": { "table": { "conditionName": "conditionQuery", ... }, ... }
```
Comment (KazuSh1geru) [inline definitions/example.js:15]: Expected Format is here.

File: definitions/example.js [hunk 2]
```diff
@@ -12,52 +12,69 @@ const commonAssertionsResult = commonAssertions({
...
-        "childTable": "third_table",
-        "childKey": "parent_id"
-      }
-    ]
+    // Format: "parentSchema": { "parentTable": [{ parentKey, childSchema, childTable, childKey }, ...], ... }
+    "dataform": {
+      "first_table": [{
+          "parentKey": "id",
+          "childSchema": "dataform",
+          "childTable": "second_table",
```
Comment (KazuSh1geru) [inline definitions/example.js:67]: add `childSchema`

File: includes/data_completeness_assertions.js [hunk 3]
```diff
@@ -11,25 +11,26 @@
...
+    const assertion = assert(`assert_data_completeness_${schemaName}_${tableName}_${columnName}`)
       .database(globalParams.database)
       .schema(globalParams.schema)
-      .description(`Check data completeness for ${tableName}.${columnName}, allowed percentage of null values: ${allowedPercentageNull}`)
+      .description(`Check data completeness for ${schemaName}.${tableName}.${columnName}, allowed percentage of null values: ${allowedPercentageNull}`)
       .tags("assert-data-completeness")
       .query(ctx => `SELECT COUNT(*) AS total_rows,
                         SUM(CASE WHEN ${columnName} IS NULL THEN 1 ELSE 0 END) AS null_count
-                        FROM ${ctx.ref(tableName)}
+                        FROM ${ctx.ref(schemaName, tableName)}
```
Comment (KazuSh1geru) [inline includes/data_completeness_assertions.js:33]: reference `schemaName.tableName`

File: includes/data_freshness_assertions.js [hunk 4]
```diff
@@ -17,19 +18,19 @@
...
+    .description(`Assert that data in ${schemaName}.${tableName} is fresh with a delay less than ${delayCondition} ${timeUnit}`)
     .tags("assert-data-freshness")
     .query(ctx => `
                 WITH
                     freshness AS (
                         SELECT
                             DATE_DIFF(CURRENT_DATE(), MAX(${dateColumn}), ${timeUnit}) AS delay
                         FROM
-                            ${ctx.ref(tableName)}
+                            ${ctx.ref(schemaName, tableName)}
```
Comment (KazuSh1geru) [inline includes/data_freshness_assertions.js:33]: reference `schemaName.tableName`

File: includes/referential_integrity_assertions.js [hunk 5]
```diff
@@ -19,17 +19,17 @@
...
+  const assertion = assert(`assert_referential_integrity_${parentSchema}_${parentTable}_${childSchema}_${childTable}`)
     .database(globalParams.database)
     .schema(globalParams.schema)
     .description(`Check referential integrity for ${childTable}.${childKey} referencing ${parentTable}.${parentKey}`)
     .tags("assert-referential-integrity")
     .query(ctx => `
           SELECT pt.${parentKey}
-          FROM ${ctx.ref(parentTable)} AS pt
-          LEFT JOIN ${ctx.ref(childTable)} AS t ON t.${childKey} = pt.${parentKey}
+          FROM ${ctx.ref(parentSchema, parentTable)} AS pt
```
Comment (KazuSh1geru) [inline includes/referential_integrity_assertions.js:31]: reference `parentSchema.parentTable`

**Comment by hrialan** [comment]
Thanks a lot for you support @KazuSh1geru !

This is indeed a super important feature that need to be implemented.

What do you think about making the schema argument optional? You can have projects with many datasets, and specify them one by one can be very long and we loose usage simplicity.

I am thinking about something like that: 

```js
const commonAssertions = require("dataform-assertions");

const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    "schema1.filter": ["Id"],
    "schema2.filter": ["Id"],
    "otherTable": ["Id"]
  }
});
```

**Comment by KazuSh1geru** [comment]
I believe my method of writing the schema in isolation is better from 2 perspectives.

As the first, the above method of specifying tables is not sufficient.
Because the way to specify in the `ref` function needs to be schema, table. `ref("schema1.filter")` will cause an error, so the proposed notation needs to be converted to `ref("schema1", "filter")` and to be correct as the ref argument.
(Parsing of strings is not recommended, as it creates complexity.)

Second, it is against the Dry Principle.
Whenever there is a table other than filter in schema1, e.g. preprocess, postprocess, etc., schema1 must be added.
If this happens in other schemas, readability will be reduced.

```
const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    "schema1.filter": ["Id"],
    "schema1.preprocess": ["Id"],
    "schema1.postprocess": ["Id"],
    "schema2.filter": ["Id"],
    "schema2.postprocess": ["Id"],
    "schema3.filter": ["Id"],
    "schema3.preprocess": ["Id"],
    "schema3.postprocess": ["Id"],
    "schema3.transform": ["Id"],
    "otherTable": ["Id"]
  }
});
```

isolation schema pattern
```
const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    "schema1": {
      "filter": ["Id"],
      "preprocess": ["Id"],
      "postprocess": ["Id"]
    },
    "schema2": {
      "filter": ["Id"],
      "postprocess": ["Id"]
    },
    "schema3": {
      "filter": ["Id"],
      "preprocess": ["Id"],
      "postprocess": ["Id"],
      "transform": ["Id"]
    },
    "otherTable": ["Id"]
  }
});
```

**Comment by KazuSh1geru** [comment]
The idea of being able to cut assertions in this package into arbitrary javascript is very nice!
I trust that this PR will be considered and better extended!
Best regards

**Comment by hrialan** [comment]
ok, I agree for the isolation pattern, It seems to be more readable.

Before merging, here are two points to improve: 

1. **[Not blocking]** Add the possibility to optionally set the schema.

Can be implemented in a future version if it is not so complicated.

As the example here for "otherTable" : 

```js
const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    "schema3": {
      "filter": ["Id"],
      "preprocess": ["Id"],
      "postprocess": ["Id"],
      "transform": ["Id"]
    },
    "otherTable": ["Id"] // THIS is not working
  }
});
```

2. **[Blocking]** How can we specify the environment in a schema name ? 

In many projects, envs are included in the dataset name. Do you have an idea on how we can do this with your solution ? 

Example:
```js
const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    "schema3" + dataform.projectConfig.vars.env : { // THIS is not working
      "filter": ["Id"],
      "preprocess": ["Id"],
      "postprocess": ["Id"],
      "transform": ["Id"]
    },
    "otherTable": ["Id"]
  }
});

Thanks!

**Comment by danilo-nzyte** [comment]
This is a requirement for us at the moment too, we have many models where names are duplicated.

> 2\. **[Blocking]** How can we specify the environment in a schema name ?

```javascript
const commonAssertionsResult = commonAssertions({
  globalAssertionsParams: {
    "schema": "dataform_assertions_" + dataform.projectConfig.vars.env,
    "location": "asia-northeast1",
    "tags": ["assertions"],
  },
  uniqueKeyConditions: {
    ["schema3" + dataform.projectConfig.vars.env] : { // Would this not solve the problem?
      "filter": ["Id"],
      "preprocess": ["Id"],
      "postprocess": ["Id"],
      "transform": ["Id"]
    },
    "otherTable": ["Id"]
  }
});```

**Comment by hrialan** [comment]
@danilo-nzyte I tested, and it worked well, I did not know that 

I will resolve the conflicts and merge the PR @KazuSh1geru 

Thanks for your help everyone !

**Pull request merged**